File: runtime/go/units/wasi/path_rename.go
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
func (w *WASI) wasi_path_rename(oldDirfd, oldPathPtr, oldPathLen, newDirfd, newPathPtr, newPathLen uint32) uint32 {
    oldRel := string(w.memory.read_string(uint64(oldPathPtr), uint64(oldPathLen)))
    // rename(2) never follows trailing symlinks: it moves the link itself and
    // replaces the destination link.
    oldHost, err := w.resolve_path(oldDirfd, oldRel, false)
    if err != wasiOk {
        return err
    }
    newRel := string(w.memory.read_string(uint64(newPathPtr), uint64(newPathLen)))
    newHost, err := w.resolve_path(newDirfd, newRel, false)
    if err != wasiOk {
        return err
    }
    // Trailing slashes (issue #42): existing non-directories were
    // ENOTDIR in resolve_path; a nonexistent slash-suffixed destination is
    // renamed bare, as wasmtime strips it — the resolved path already is.
    // syscall.Rename, not os.Rename: Go's os.Rename wrapper Lstats the
    // destination and, when it is a directory, returns a synthetic EEXIST on
    // macOS instead of letting rename(2) replace an empty target dir — the
    // atomic dir-onto-empty-dir semantics the suite requires. The raw syscall
    // has the correct POSIX behaviour (ENOTEMPTY on a non-empty target,
    // EISDIR/ENOTDIR on type mismatches).
    if e := syscall.Rename(oldHost, newHost); e != nil {
        return w.fs_errno(e)
    }
    return wasiOk
}

File: runtime/java/units/wasi/path_rename.java
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
int wasi_path_rename(int oldDirfd, int oldPathPtr, int oldPathLen, int newDirfd, int newPathPtr,
                     int newPathLen) {
    String oldRel = new String(
        memory.read_string(Integer.toUnsignedLong(oldPathPtr), Integer.toUnsignedLong(oldPathLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    // rename(2) never follows trailing symlinks: it moves the link itself and
    // replaces the destination link.
    Resolved oldR = resolve_path(oldDirfd, oldRel, false);
    if (oldR.errno != WASI_OK) {
        return oldR.errno;
    }
    String newRel = new String(
        memory.read_string(Integer.toUnsignedLong(newPathPtr), Integer.toUnsignedLong(newPathLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    Resolved newR = resolve_path(newDirfd, newRel, false);
    if (newR.errno != WASI_OK) {
        return newR.errno;
    }
    java.nio.file.Path oldP = java.nio.file.Paths.get(oldR.path);
    java.nio.file.Path newP = java.nio.file.Paths.get(newR.path);
    // Trailing slashes (issue #42): existing non-directories were
    // ENOTDIR in resolve_path; a nonexistent slash-suffixed destination is
    // renamed bare, as wasmtime strips it — the normalized Path already is.
    // rename(2) reports type mismatches between the endpoints with specific
    // errnos that Java's generic FileSystemException flattens to EIO, so
    // pre-check them: renaming a directory onto an existing non-directory is
    // ENOTDIR, and a non-directory onto an existing directory is EISDIR. The
    // matching directory-onto-nonempty-directory case surfaces as
    // DirectoryNotEmptyException -> ENOTEMPTY through fs_errno on its own.
    if (java.nio.file.Files.exists(newP, java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
        boolean oldIsDir = java.nio.file.Files.isDirectory(oldP, java.nio.file.LinkOption.NOFOLLOW_LINKS);
        boolean newIsDir = java.nio.file.Files.isDirectory(newP, java.nio.file.LinkOption.NOFOLLOW_LINKS);
        if (oldIsDir && !newIsDir) {
            return WASI_NOTDIR;
        }
        if (!oldIsDir && newIsDir) {
            return WASI_ISDIR;
        }
    }
    try {
        java.nio.file.Files.move(oldP, newP, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
    } catch (java.io.IOException ex) {
        return fs_errno(ex);
    }
    return WASI_OK;
}
